syntax = "proto3";

package transport_catalogue_serialize;

import "graph.proto";

message RoutingSettings {
    double bus_velocity = 1;
    uint32 bus_wait_time = 2;
}

message StopItemPair { 
    uint64 wait_on_stop_id = 1;
    uint64 stop_id = 2;
}

message StopnameAndStopItemPairPair {
    string key = 1;
    StopItemPair value = 2; 
}

message WaitItem {
    string stop_name = 1;
    double time = 2;
}

message BusItem {
    string bus_name = 1;
    uint32 span_count = 2;
    double time = 3;
}

message RouteItem {
    oneof variant {
        WaitItem wait_item = 1;
        BusItem bus_item = 2;
    }
}

message EdgeIdAndRouteItemPair {
    uint64 key = 1;
    RouteItem value = 2;
}

message Router {
    RoutingSettings routing_settings = 1;
    Graph graph = 2; //n^2+n
    repeated StopnameAndStopItemPairPair stopname_to_vertex_id_pair = 3; //m
    repeated EdgeIdAndRouteItemPair edge_id_to_route_item = 4; //m
}
